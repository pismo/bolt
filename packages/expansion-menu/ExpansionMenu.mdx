---
'name': ExpansionMenu
---

import { useState } from 'react'
import { Playground } from 'docz'
import { Bolt } from '@pismo/bolt-core'
import { ExpansionMenu } from './ExpansionMenu'

# ExpansionMenu

**ExpansionMenu** fornece um Panel list com ferramentas para mostrar um menu de com efeito de abrir e fechar

### Propriedades

| **prop**        | **type**                                        | **required** | **desc**                                                                               |
| --------------- | ----------------------------------------------- | :----------: | -------------------------------------------------------------------------------------- |
| data            | {[key: string]: any}[]                          |     true     | array de objetos que formará o menu                                                    |
| getComponent    | (item: {[key: string]: any}) => React.ReactNode |     true     | método que constroi o componente unitário da lista - item é um objeto unitário do data |
| SumaryComponent | () => React.ReactNode                           |     true     | Componente que formará o sumário                                                       |
| onChange        | (value: {[key: string]: any} ) => void          |    false     | método que recebe o valor selecionado                                                  |

### Exemplo

```javascript
import { Bolt } from '@pismo/bolt-core'
import { ExpansionMenu, FLAG } from '@pismo/bolt-flag-panel'
```

<Playground>
    {
        () => {
            const data = [
                {description: 'Português'},
                {description: 'Inglês'},
                {description: 'Espanhol'}
            ]
            const [current, setCurrent] = useState(data[0])

            const handleChange = (value) => {
                setCurrent(value)
            }

            return (
                <Bolt>
                    <ExpansionMenu
                        data={data}
                        getComponent={(item) => <div>{item.description}</div>}
                        SumaryComponent={() => <div>Languages => {current.description}</div>}
                        onChange={handleChange}
                    />
                </Bolt>
            )
        }
    }

</Playground>
