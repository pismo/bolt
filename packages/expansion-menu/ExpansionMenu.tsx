import * as React from 'react'

import ExpansionPanel from '@material-ui/core/ExpansionPanel'
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary'
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails'
import ExpandMoreIcon from '@material-ui/icons/ExpandMore'
import MenuItem from '@material-ui/core/MenuItem'

const { useState } = React
interface ExpansionMenuProps {
  data: { [key: string]: any }[]
  getComponent: (item: { [key: string]: any }) => React.ReactNode
  SumaryComponent: React.ReactType
  onChange?: (item: { [key: string]: any }) => void
  expansionMenuClasses?: any
  sumaryClasses?: any
  menuItemClasses?: any
  expandMoreIconClasses?: any
}

const ExpansionMenu: React.FC<ExpansionMenuProps> = ({
  data,
  getComponent,
  SumaryComponent,
  onChange,
  expansionMenuClasses,
  sumaryClasses,
  menuItemClasses,
  expandMoreIconClasses
}: ExpansionMenuProps) => {
  const [isExpanded, setIsExpanded] = useState<boolean>(false)

  const toggleExpanded = () => {
    setIsExpanded(!isExpanded)
  }

  const handleSelect = item => () => {
    setIsExpanded(false)
    if (onChange) {
      onChange(item)
    }
  }

  return (
    <ExpansionPanel
      expanded={isExpanded}
      square={true}
      onChange={toggleExpanded}
      className={`ExpansionMenu ${expansionMenuClasses}`}
    >
      <ExpansionPanelSummary
        className={`ExpansionMenu-summary ${sumaryClasses}`}
        expandIcon={
          <ExpandMoreIcon
            className={`ExpansionMenu-expandMoreIcon ${expandMoreIconClasses}`}
          />
        }
      >
        <SumaryComponent />
      </ExpansionPanelSummary>
      <ExpansionPanelDetails>
        {data.map((item, index) => (
          <MenuItem
            className={`ExpansionMenu-menuItem ${menuItemClasses}`}
            key={index}
            onClick={handleSelect(item)}
          >
            {getComponent(item)}
          </MenuItem>
        ))}
      </ExpansionPanelDetails>
    </ExpansionPanel>
  )
}

export { ExpansionMenu }
