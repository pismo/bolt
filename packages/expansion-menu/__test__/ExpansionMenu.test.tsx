import {
  cleanup,
  render,
  fireEvent,
  wait
  // getByText,
  // waitForDomChange
} from '@testing-library/react'
import 'jest-dom/extend-expect'
import * as React from 'react'

import { ExpansionMenu } from '../ExpansionMenu'
import { Bolt } from '@pismo/bolt-core'

const { useState } = React

afterEach(cleanup)

const Component: React.FC<{ change?: (value: any) => void }> = ({ change }) => {
  const data = [
    { description: 'Português' },
    { description: 'Inglês' },
    { description: 'Espanhol' }
  ]

  const [current] = useState(data[0])

  const handleChange = value => {
    if (change) change(value)
  }

  return (
    <Bolt>
      <ExpansionMenu
        data={data}
        getComponent={item => (
          <div data-testid={item.description}>{item.description}</div>
        )}
        SumaryComponent={() => <div>Languages => {current.description}</div>}
        onChange={handleChange}
      />
    </Bolt>
  )
}

describe('@pismo/bolt-expansion-menu', () => {
  test('it should display the components correctly', () => {
    const { getByText } = render(<Component />)

    expect(getByText('Languages => Português')).toBeDefined()
    expect(getByText('Português')).toBeDefined()
    expect(getByText('Inglês')).toBeDefined()
    expect(getByText('Espanhol')).toBeDefined()
  })

  test('it should return the chosen value', () => {
    const fn = jest.fn(value => value.description)

    const { getByTestId } = render(<Component change={fn} />)

    const usBt = getByTestId('Inglês')
    const ptBt = getByTestId('Português')
    const esBt = getByTestId('Espanhol')

    fireEvent.click(usBt)

    wait(() => true, { timeout: 1000 }).then(() => {
      expect(fn).toBeCalledTimes(1)
      expect(fn).toHaveReturnedWith('Inglês')
    })

    fireEvent.click(ptBt)

    wait(() => true, { timeout: 1000 }).then(() => {
      expect(fn).toBeCalledTimes(2)
      expect(fn).toReturnWith('Português')
    })

    fireEvent.click(esBt)

    wait(() => true, { timeout: 1000 }).then(() => {
      expect(fn).toBeCalledTimes(3)
      expect(fn).toReturnWith('Espanhol')
    })
  })
})
