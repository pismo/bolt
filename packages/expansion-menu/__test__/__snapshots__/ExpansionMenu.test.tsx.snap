// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@pismo/bolt-expansion-menu it should display the components correctly 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiExpansionPanel-root ExpansionMenu"
  >
    <div
      aria-disabled="false"
      aria-expanded="false"
      class="MuiButtonBase-root MuiExpansionPanelSummary-root ExpansionMenu-summary"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiExpansionPanelSummary-content"
      >
        <div>
          Languages =&gt; 
          Português
        </div>
      </div>
      <div
        aria-disabled="false"
        aria-hidden="true"
        class="MuiButtonBase-root MuiIconButton-root MuiExpansionPanelSummary-expandIcon MuiIconButton-edgeEnd"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root ExpansionMenu-expandMoreIcon"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
    <div
      class="MuiCollapse-container MuiCollapse-hidden"
      style="min-height: 0px;"
    >
      <div
        class="MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              class="MuiExpansionPanelDetails-root"
            >
              <li
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
                role="menuitem"
                tabindex="-1"
              >
                <div
                  data-testid="Português"
                >
                  Português
                </div>
                <span
                  class="MuiTouchRipple-root"
                />
              </li>
              <li
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
                role="menuitem"
                tabindex="-1"
              >
                <div
                  data-testid="Inglês"
                >
                  Inglês
                </div>
                <span
                  class="MuiTouchRipple-root"
                />
              </li>
              <li
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
                role="menuitem"
                tabindex="-1"
              >
                <div
                  data-testid="Espanhol"
                >
                  Espanhol
                </div>
                <span
                  class="MuiTouchRipple-root"
                />
              </li>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
